code/datums/chat_message.dm

```diff
@@ -0,0 +1,251 @@
+	if (gcDestroyed)
+		return
+	animate(message, alpha = 0, time = fadetime, flags = ANIMATION_PARALLEL)
+	spawn(fadetime)
+		qdel(src)
+
+/**
+  * Creates a message overlay at a defined location for a given speaker
+  *
+  * Arguments:
+  * * speaker - The atom who is saying this message
+  * * message_language - The language that the message is said in
+  * * raw_message - The text content of the message
+  * * spans - Additional classes to be added to the message
+  * * message_mode - Bitflags relating to the mode of the message
+  */
+/mob/proc/create_chat_message(atom/movable/speaker, datum/language/message_language, raw_message, mode)
+	var/extra_classes = list()
+
+	// Check for virtual speakers (aka hearing a message through a radio)
+	if (istype(speaker, /atom/movable/virtualspeaker))
+		return
+
+	if (mode == SPEECH_MODE_WHISPER)
+		extra_classes |= "small"
+
+	if (client.toggle_runechat_outlines)
+		extra_classes |= "black_outline"
+
+	var/dist = get_dist(src, speaker)
+	switch (dist)
+		if (4 to 5)
+			extra_classes |= "small"
+		if (5 to 16)
+			extra_classes |= "very_small"
+
+	if (!say_understands(speaker, message_language))
+		raw_message = message_language.scramble(raw_message)
+
+	// Display visual above source
+	new /datum/chatmessage(raw_message, speaker, src, extra_classes)
+
+// Tweak these defines to change the available color ranges
+#define CM_COLOR_SAT_MIN	0.6
+#define CM_COLOR_SAT_MAX	0.7
+#define CM_COLOR_LUM_MIN	0.65
+#define CM_COLOR_LUM_MAX	0.75
+
+/**
+  * Gets a color for a name, will return the same color for a given string consistently within a round.atom
+  *
+  * Note that this proc aims to produce pastel-ish colors using the HSL colorspace. These seem to be favorable for displaying on the map.
+  *
+  * Arguments:
+  * * name - The name to generate a color for
+  * * sat_shift - A value between 0 and 1 that will be multiplied against the saturation
+  * * lum_shift - A value between 0 and 1 that will be multiplied against the luminescence
+  */
+/datum/chatmessage/proc/colorize_string(name, sat_shift = 1, lum_shift = 1)
+	// seed to help randomness
+	var/static/rseed = rand(1,26)
+
+	// get hsl using the selected 6 characters of the md5 hash
+	var/hash = copytext(md5(name + "[world_startup_time]"), rseed, rseed + 6)
+	var/h = hex2num(copytext(hash, 1, 3)) * (360 / 255)
+	var/s = (hex2num(copytext(hash, 3, 5)) >> 2) * ((CM_COLOR_SAT_MAX - CM_COLOR_SAT_MIN) / 63) + CM_COLOR_SAT_MIN
+	var/l = (hex2num(copytext(hash, 5, 7)) >> 2) * ((CM_COLOR_LUM_MAX - CM_COLOR_LUM_MIN) / 63) + CM_COLOR_LUM_MIN
+
+	// adjust for shifts
+	s *= clamp(sat_shift, 0, 1)
+	l *= clamp(lum_shift, 0, 1)
+
+	// convert to rgba
+	var/h_int = round(h/60) // mapping each section of H to 60 degree sections
+	var/c = (1 - abs(2 * l - 1)) * s
+	var/x = c * (1 - abs((h / 60) % 2 - 1))
+	var/m = l - c * 0.5
+	x = (x + m) * 255
+	c = (c + m) * 255
+	m *= 255
+	switch(h_int)
+		if(0)
+			return "#[num2hex(c, 2)][num2hex(x, 2)][num2hex(m, 2)]C8"
```

Comment:
you should be able to replace these with 
```suggestion
			return "#[rgb(c, x, m)]C8"
```
I remember testing this a while back and it was a lot faster